ros/src/twist_controller/twist_controller_mod.py

```diff
@@ -32,60 +32,55 @@ def __init__(self, default_update_interval, wheel_base, steer_ratio, min_speed,
     def handle_dynamic_variable_update(self, config, level):
         # reset PID controller to use new parameters
         self.setup_pid_controllers(config['dyn_velo_proportional_control'], config[
-            'dyn_accel_proportional_control'], config['dyn_accel_integral_control'], self.max_deceleration, self.max_acceleration)
-
-        # reset acceleration cruise constant
-        self.acceleration_cruise_constant = config['dyn_acceleration_cruise_constant']
+            'dyn_velo_integral_control'], config['dyn_braking_proportional_control'], config['dyn_braking_integral_control'])
 
         return config
 
-    def setup_pid_controllers(self, velo_p, accel_p, accel_i, max_deceleration, max_acceleration):
-        rospy.loginfo("Initializing PID controllers with velo_P: {}, accel_P: {}, accel_I: {}"
-                      .format(velo_p, accel_p, accel_i))
+    def setup_pid_controllers(self, velo_p, velo_i, braking_p, braking_i):
+        rospy.loginfo("Initializing PID controllers with velo_P: {}, velo_I: {}, braking_P: {}, braking_I: {}"
+                      .format(velo_p, velo_i, braking_p, braking_i))
         # create velocity pid controller thresholded between min and max
         # acceleration values
         self.velocity_pid_controller = PID(
-            velo_p, 0, 0, max_deceleration, max_acceleration)
+            velo_p, velo_i, 0, 0, 1)
 
         # create acceleration pid controller thresholded between 0% and 100%
         # for throttle
-        self.acceleration_pid_controller = PID(accel_p, accel_i, 0.0, 0.0, 1.0)
+        self.braking_pid_controller = PID(
+            braking_p, braking_i, 0.0, 0.0, 10000)
 
     def control(self, target_linear_velocity, target_angular_velocity, current_linear_velocity):
         # compute timestep
         timestep = self.compute_timestep()
-        # use velocity pid controller to compute desired acceleration
-        acceleration_command = self.compute_acceleration_command(
-            target_linear_velocity - current_linear_velocity, timestep)
-        rospy.logwarn('Current linear velocity %.2f, target linear velocity %.2f, desired total acceleration %.2f',
-                      current_linear_velocity, target_linear_velocity, acceleration_command)
+        velocity_error = target_linear_velocity - current_linear_velocity
+
+        if (target_angular_velocity == 0 and current_linear_velocity == 0):
+            # reset integrators if we're at a stop
+            self.reset()
+
+        # use throttle if we want to speed up or if we want to slow down just slightly
+        throttle_command = self.velocity_pid_controller.step(
+            velocity_error, timestep) if velocity_error > 0 or velocity_error > (-0.075 * current_linear_velocity) else 0
```

Comment:
Should the -0.075 (-7.5%) for throttle and -0.1 (-10%) be named parameters?  Also, to guarantee the braking activates outside of the throttle area when tuning, maybe the brake parameter should be set by the throttle parameter minus a margin (2.5%).

Also, I think to avoid activation hunting as it crosses the threshold, we could apply some hysteresis to the threshold parameters, like when throttle > 0 then throttle param = 7.5% but when throttle == 0 then throttle param = 2.5%.  If brake parameter is a margin offset from throttle, then it would also automatically adjust.

Reply:
Yep, these should indeed be named parameters. I meant to do that and then slipped my mind before I commited up to github.

Sure, it makes sense to automatically set the braking parameter as a dependent variable of the throttle parameter. I had been adjusting them in pairs manually, but that would save me a step.

I did adjust the throttle param based on the desire to accelerate or decelerate, and it did track the target speed slightly better during deceleration. Thanks for the suggestion.